File: backend/app/api/v1/schemas/admin_products.py
"""Admin product API schemas — camelCase for FE (Phase H)."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

ProductStatusLiteral = Literal["draft", "published", "archived"]


class CamelModel(BaseModel):
    model_config = ConfigDict(
        populate_by_name=True,
        ser_json_by_alias=True,
    )


class AdminProductSpecIn(CamelModel):
    label: str = Field(min_length=1, max_length=128)
    value: str = Field(min_length=1, max_length=2000)

    @field_validator("label", "value")
    @classmethod
    def strip_text(cls, v: str) -> str:
        s = v.strip()
        if not s:
            raise ValueError("must not be blank")
        return s


class AdminProductSpecOut(CamelModel):
    label: str
    value: str
    sort_order: int = Field(serialization_alias="sortOrder")


class AdminProductImageIn(CamelModel):
    url: str = Field(min_length=1, max_length=1024)

    @field_validator("url")
    @classmethod
    def strip_url(cls, v: str) -> str:
        s = v.strip()
        if not s:
            raise ValueError("url must not be blank")
        return s


class AdminProductImageOut(CamelModel):
    url: str
    sort_order: int = Field(serialization_alias="sortOrder")


class AdminProductListItemOut(CamelModel):
    id: str
    slug: str
    name: str
    status: str
    price: float
    currency: str
    image_url: str = Field(serialization_alias="imageUrl")
    in_stock: bool = Field(serialization_alias="inStock")
    featured: bool
    category_slugs: list[str] = Field(serialization_alias="categorySlugs")
    published_at: str | None = Field(default=None, serialization_alias="publishedAt")
    updated_at: str = Field(serialization_alias="updatedAt")
    created_at: str = Field(serialization_alias="createdAt")


class AdminProductListOut(CamelModel):
    items: list[AdminProductListItemOut]
    page: int
    page_size: int = Field(serialization_alias="pageSize")
    total_items: int = Field(serialization_alias="totalItems")
    total_pages: int = Field(serialization_alias="totalPages")
    has_next: bool = Field(serialization_alias="hasNext")
    has_prev: bool = Field(serialization_alias="hasPrev")


class AdminProductDetailOut(CamelModel):
    id: str
    slug: str
    name: str
    status: str
    short_description: str = Field(serialization_alias="shortDescription")
    description: str
    price: float
    compare_at_price: float | None = Field(
        default=None,
        serialization_alias="compareAtPrice",
    )
    currency: str
    image_url: str = Field(serialization_alias="imageUrl")
    image_alt: str = Field(serialization_alias="imageAlt")
    images: list[AdminProductImageOut]
    category_slugs: list[str] = Field(serialization_alias="categorySlugs")
    category_label: str | None = Field(default=None, serialization_alias="categoryLabel")
    badges: list[str]
    features: list[str]
    specs: list[AdminProductSpecOut]
    in_stock: bool = Field(serialization_alias="inStock")
    featured: bool
    max_quantity_per_order: int = Field(serialization_alias="maxQuantityPerOrder")
    purchase_limit_per_customer: int | None = Field(
        default=None,
        serialization_alias="purchaseLimitPerCustomer",
    )
    seo_title: str | None = Field(default=None, serialization_alias="seoTitle")
    seo_description: str | None = Field(default=None, serialization_alias="seoDescription")
    rating: float
    review_count: int = Field(serialization_alias="reviewCount")
    published_at: str | None = Field(default=None, serialization_alias="publishedAt")
    created_at: str = Field(serialization_alias="createdAt")
    updated_at: str = Field(serialization_alias="updatedAt")


class AdminProductCreateIn(CamelModel):
    """Create product. Default status = draft. Optional id acts as SKU-like primary key."""

    model_config = ConfigDict(populate_by_name=True)

    id: str | None = Field(
        default=None,
        min_length=3,
        max_length=64,
        description="Optional stable product id (SKU-like). Auto-generated if omitted.",
    )
    slug: str = Field(min_length=2, max_length=255)
    name: str = Field(min_length=1, max_length=255)
    short_description: str = Field(
        default="",
        max_length=2000,
        validation_alias="shortDescription",
    )
    description: str = Field(default="", max_length=20000)
    price: float = Field(gt=0, description="Unit price in major currency units (USD)")
    compare_at_price: float | None = Field(
        default=None,
        validation_alias="compareAtPrice",
        gt=0,
    )
    currency: str = Field(default="USD", min_length=3, max_length=3)
    image_url: str = Field(default="", max_length=1024, validation_alias="imageUrl")
    image_alt: str = Field(default="", max_length=512, validation_alias="imageAlt")
    images: list[AdminProductImageIn] = Field(default_factory=list)
    category_slugs: list[str] = Field(
        default_factory=list,
        validation_alias="categorySlugs",
    )
    category_label: str | None = Field(
        default=None,
        max_length=128,
        validation_alias="categoryLabel",
    )
    badges: list[str] = Field(default_factory=list)
    features: list[str] = Field(default_factory=list)
    specs: list[AdminProductSpecIn] = Field(default_factory=list)
    in_stock: bool = Field(default=True, validation_alias="inStock")
    featured: bool = False
    status: ProductStatusLiteral = "draft"
    max_quantity_per_order: int = Field(
        default=9,
        ge=1,
        le=99,
        validation_alias="maxQuantityPerOrder",
    )
    purchase_limit_per_customer: int | None = Field(
        default=None,
        ge=1,
        le=999,
        validation_alias="purchaseLimitPerCustomer",
    )
    seo_title: str | None = Field(default=None, max_length=255, validation_alias="seoTitle")
    seo_description: str | None = Field(
        default=None,
        max_length=2000,
        validation_alias="seoDescription",
    )

    @field_validator("slug")
    @classmethod
    def normalize_slug(cls, v: str) -> str:
        s = v.strip().lower()
        if not s:
            raise ValueError("slug must not be blank")
        return s

    @field_validator("id")
    @classmethod
    def normalize_id(cls, v: str | None) -> str | None:
        if v is None:
            return None
        s = v.strip()
        return s or None

    @field_validator("currency")
    @classmethod
    def upper_currency(cls, v: str) -> str:
        return v.strip().upper() or "USD"

    @field_validator("category_slugs")
    @classmethod
    def clean_slugs(cls, v: list[str]) -> list[str]:
        out: list[str] = []
        seen: set[str] = set()
        for raw in v:
            s = raw.strip().lower()
            if not s or s in seen:
                continue
            seen.add(s)
            out.append(s)
        return out


class AdminProductUpdateIn(CamelModel):
    """Partial update. Omit fields to leave unchanged. Nested images/specs replace when set."""

    model_config = ConfigDict(populate_by_name=True)

    slug: str | None = Field(default=None, min_length=2, max_length=255)
    name: str | None = Field(default=None, min_length=1, max_length=255)
    short_description: str | None = Field(
        default=None,
        max_length=2000,
        validation_alias="shortDescription",
    )
    description: str | None = Field(default=None, max_length=20000)
    price: float | None = Field(default=None, gt=0)
    compare_at_price: float | None = Field(
        default=None,
        validation_alias="compareAtPrice",
    )
    image_url: str | None = Field(default=None, max_length=1024, validation_alias="imageUrl")
    image_alt: str | None = Field(default=None, max_length=512, validation_alias="imageAlt")
    images: list[AdminProductImageIn] | None = None
    category_slugs: list[str] | None = Field(default=None, validation_alias="categorySlugs")
    category_label: str | None = Field(
        default=None,
        max_length=128,
        validation_alias="categoryLabel",
    )
    badges: list[str] | None = None
    features: list[str] | None = None
    specs: list[AdminProductSpecIn] | None = None
    in_stock: bool | None = Field(default=None, validation_alias="inStock")
    featured: bool | None = None
    status: ProductStatusLiteral | None = None
    max_quantity_per_order: int | None = Field(
        default=None,
        ge=1,
        le=99,
        validation_alias="maxQuantityPerOrder",
    )
    purchase_limit_per_customer: int | None = Field(
        default=None,
        validation_alias="purchaseLimitPerCustomer",
    )
    seo_title: str | None = Field(default=None, max_length=255, validation_alias="seoTitle")
    seo_description: str | None = Field(
        default=None,
        max_length=2000,
        validation_alias="seoDescription",
    )

    @field_validator("slug")
    @classmethod
    def normalize_slug(cls, v: str | None) -> str | None:
        if v is None:
            return None
        s = v.strip().lower()
        if not s:
            raise ValueError("slug must not be blank")
        return s

    @field_validator("category_slugs")
    @classmethod
    def clean_slugs(cls, v: list[str] | None) -> list[str] | None:
        if v is None:
            return None
        out: list[str] = []
        seen: set[str] = set()
        for raw in v:
            s = raw.strip().lower()
            if not s or s in seen:
                continue
            seen.add(s)
            out.append(s)
        return out


class AdminProductStatusOut(CamelModel):
    """Publish / unpublish response (full detail for FE convenience)."""

    product: AdminProductDetailOut

File: src/lib/api/admin-products.ts
/**
 * Admin products API (Phase H).
 * Browser-only: credentials: "include" for FastAPI HttpOnly cookies.
 */

import { getApiBaseUrl } from "@/lib/api/config";

export type AdminProductStatus = "draft" | "published" | "archived";

export type AdminProductSpec = {
  label: string;
  value: string;
  sortOrder?: number;
};

export type AdminProductImage = {
  url: string;
  sortOrder?: number;
};

export type AdminProductListItem = {
  id: string;
  slug: string;
  name: string;
  status: AdminProductStatus | string;
  price: number;
  currency: string;
  imageUrl: string;
  /** Internal supplier page; returned only by authenticated Admin endpoints. */
  supplierUrl: string;
  inStock: boolean;
  /** Available units (Phase L). 0 blocks checkout. */
  stockQty?: number;
  featured: boolean;
  categorySlugs: string[];
  publishedAt: string | null;
  createdAt: string;
  updatedAt: string;
};

export type AdminProductDetail = AdminProductListItem & {
  shortDescription: string;
  description: string;
  compareAtPrice: number | null;
  imageAlt: string;
  supplierUrl: string;
  images: AdminProductImage[];
  categoryLabel: string | null;
  badges: string[];
  features: string[];
  specs: AdminProductSpec[];
  maxQuantityPerOrder: number;
  purchaseLimitPerCustomer: number | null;
  seoTitle: string | null;
  seoDescription: string | null;
  rating: number;
  reviewCount: number;
};

export type AdminProductListResponse = {
  items: AdminProductListItem[];
  page: number;
  pageSize: number;
  totalItems: number;
  totalPages: number;
  hasNext: boolean;
  hasPrev: boolean;
};

export type AdminProductCreateInput = {
  id?: string;
  slug: string;
  name: string;
  shortDescription?: string;
  description?: string;
  price: number;
  compareAtPrice?: number | null;
  currency?: string;
  imageUrl?: string;
  imageAlt?: string;
  supplierUrl?: string;
  images?: { url: string }[];
  categorySlugs?: string[];
  categoryLabel?: string | null;
  badges?: string[];
  features?: string[];
  specs?: { label: string; value: string }[];
  inStock?: boolean;
  stockQty?: number;
  featured?: boolean;
  status?: AdminProductStatus;
  maxQuantityPerOrder?: number;
  purchaseLimitPerCustomer?: number | null;
  seoTitle?: string | null;
  seoDescription?: string | null;
};

export type AdminProductUpdateInput = Partial<AdminProductCreateInput>;

export class AdminProductsApiError extends Error {
  status: number;
  code?: string;

  constructor(message: string, status: number, code?: string) {
    super(message);
    this.name = "AdminProductsApiError";
    this.status = status;
    this.code = code;
  }
}

function apiUrl(path: string): string {
  const base = getApiBaseUrl();
  const p = path.startsWith("/") ? path : `/${path}`;
  return `${base}${p}`;
}

async function parseJson(res: Response): Promise<Record<string, unknown>> {
  const text = await res.text();
  if (!text) return {};
  try {
    return JSON.parse(text) as Record<string, unknown>;
  } catch {
    throw new AdminProductsApiError(
      `Invalid JSON from admin products API (${res.status})`,
      res.status,
    );
  }
}

function errorFromResponse(
  res: Response,
  data: Record<string, unknown>,
): AdminProductsApiError {
  const detail = data.detail;
  if (typeof detail === "string") {
    return new AdminProductsApiError(detail, res.status);
  }
  if (detail && typeof detail === "object") {
    const d = detail as Record<string, unknown>;
    const msg =
      typeof d.message === "string"
        ? d.message
        : typeof d.detail === "string"
          ? d.detail
          : `HTTP ${res.status}`;
    const code = typeof d.code === "string" ? d.code : undefined;
    return new AdminProductsApiError(msg, res.status, code);
  }
  // FastAPI 422 validation errors
  if (Array.isArray(detail)) {
    const first = detail[0] as { msg?: string } | undefined;
    return new AdminProductsApiError(
      first?.msg || `Validation error (${res.status})`,
      res.status,
      "validation_error",
    );
  }
  return new AdminProductsApiError(`HTTP ${res.status}`, res.status);
}

function asNumber(v: unknown, fallback = 0): number {
  return typeof v === "number" && Number.isFinite(v) ? v : fallback;
}

function asString(v: unknown, fallback = ""): string {
  return typeof v === "string" ? v : fallback;
}

function asStringArray(v: unknown): string[] {
  if (!Array.isArray(v)) return [];
  return v.filter((x): x is string => typeof x === "string");
}

function mapListItem(raw: Record<string, unknown>): AdminProductListItem {
  return {
    id: asString(raw.id),
    slug: asString(raw.slug),
    name: asString(raw.name),
    status: asString(raw.status),
    price: asNumber(raw.price),
    currency: asString(raw.currency, "USD"),
    imageUrl: asString(raw.imageUrl ?? raw.image_url),
    supplierUrl: asString(raw.supplierUrl ?? raw.supplier_url),
    inStock: Boolean(raw.inStock ?? raw.in_stock),
    stockQty:
      raw.stockQty != null || raw.stock_qty != null
        ? asNumber(raw.stockQty ?? raw.stock_qty)
        : undefined,
    featured: Boolean(raw.featured),
    categorySlugs: asStringArray(raw.categorySlugs ?? raw.category_slugs),
    publishedAt:
      raw.publishedAt != null || raw.published_at != null
        ? asString(raw.publishedAt ?? raw.published_at)
        : null,
    createdAt: asString(raw.createdAt ?? raw.created_at),
    updatedAt: asString(raw.updatedAt ?? raw.updated_at),
  };
}

function mapDetail(raw: Record<string, unknown>): AdminProductDetail {
  const imagesRaw = Array.isArray(raw.images) ? raw.images : [];
  const specsRaw = Array.isArray(raw.specs) ? raw.specs : [];
  const compareRaw = raw.compareAtPrice ?? raw.compare_at_price;

  return {
    ...mapListItem(raw),
    shortDescription: asString(raw.shortDescription ?? raw.short_description),
    description: asString(raw.description),
    compareAtPrice:
      compareRaw == null || compareRaw === ""
        ? null
        : asNumber(compareRaw),
    imageAlt: asString(raw.imageAlt ?? raw.image_alt),
    supplierUrl: asString(raw.supplierUrl ?? raw.supplier_url),
    images: imagesRaw.map((img) => {
      const i = img as Record<string, unknown>;
      return {
        url: asString(i.url),
        sortOrder: asNumber(i.sortOrder ?? i.sort_order),
      };
    }),
    categoryLabel:
      raw.categoryLabel != null || raw.category_label != null
        ? asString(raw.categoryLabel ?? raw.category_label)
        : null,
    badges: asStringArray(raw.badges),
    features: asStringArray(raw.features),
    specs: specsRaw.map((s) => {
      const row = s as Record<string, unknown>;
      return {
        label: asString(row.label),
        value: asString(row.value),
        sortOrder: asNumber(row.sortOrder ?? row.sort_order),
      };
    }),
    maxQuantityPerOrder: asNumber(
      raw.maxQuantityPerOrder ?? raw.max_quantity_per_order,
      9,
    ),
    purchaseLimitPerCustomer:
      raw.purchaseLimitPerCustomer != null ||
      raw.purchase_limit_per_customer != null
        ? asNumber(
            raw.purchaseLimitPerCustomer ?? raw.purchase_limit_per_customer,
          )
        : null,
    seoTitle:
      raw.seoTitle != null || raw.seo_title != null
        ? asString(raw.seoTitle ?? raw.seo_title)
        : null,
    seoDescription:
      raw.seoDescription != null || raw.seo_description != null
        ? asString(raw.seoDescription ?? raw.seo_description)
        : null,
    rating: asNumber(raw.rating),
    reviewCount: asNumber(raw.reviewCount ?? raw.review_count),
  };
}

export type ListAdminProductsParams = {
  status?: string;
  q?: string;
  page?: number;
  pageSize?: number;
};

/** GET /api/v1/admin/products */
export async function listAdminProducts(
  params: ListAdminProductsParams = {},
): Promise<AdminProductListResponse> {
  const qs = new URLSearchParams();
  if (params.status) qs.set("status", params.status);
  if (params.q) qs.set("q", params.q);
  qs.set("page", String(params.page ?? 1));
  qs.set("pageSize", String(params.pageSize ?? 20));

  const res = await fetch(apiUrl(`/api/v1/admin/products?${qs.toString()}`), {
    method: "GET",
    credentials: "include",
    headers: { Accept: "application/json" },
  });
  const data = await parseJson(res);
  if (!res.ok) throw errorFromResponse(res, data);

  const itemsRaw = Array.isArray(data.items) ? data.items : [];
  return {
    items: itemsRaw.map((i) => mapListItem(i as Record<string, unknown>)),
    page: asNumber(data.page, 1),
    pageSize: asNumber(data.pageSize ?? data.page_size, 20),
    totalItems: asNumber(data.totalItems ?? data.total_items),
    totalPages: asNumber(data.totalPages ?? data.total_pages, 1),
    hasNext: Boolean(data.hasNext ?? data.has_next),
    hasPrev: Boolean(data.hasPrev ?? data.has_prev),
  };
}

/** GET /api/v1/admin/products/{id} */
export async function getAdminProduct(id: string): Promise<AdminProductDetail> {
  const res = await fetch(
    apiUrl(`/api/v1/admin/products/${encodeURIComponent(id)}`),
    {
      method: "GET",
      credentials: "include",
      headers: { Accept: "application/json" },
    },
  );
  const data = await parseJson(res);
  if (!res.ok) throw errorFromResponse(res, data);
  return mapDetail(data);
}

/** POST /api/v1/admin/products */
export async function createAdminProduct(
  body: AdminProductCreateInput,
): Promise<AdminProductDetail> {
  const res = await fetch(apiUrl("/api/v1/admin/products"), {
    method: "POST",
    credentials: "include",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
    body: JSON.stringify(body),
  });
  const data = await parseJson(res);
  if (!res.ok) throw errorFromResponse(res, data);
  return mapDetail(data);
}

/** PATCH /api/v1/admin/products/{id} */
export async function updateAdminProduct(
  id: string,
  body: AdminProductUpdateInput,
): Promise<AdminProductDetail> {
  const res = await fetch(
    apiUrl(`/api/v1/admin/products/${encodeURIComponent(id)}`),
    {
      method: "PATCH",
      credentials: "include",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    },
  );
  const data = await parseJson(res);
  if (!res.ok) throw errorFromResponse(res, data);
  return mapDetail(data);
}

/** POST /api/v1/admin/products/{id}/publish */
export async function publishAdminProduct(
  id: string,
): Promise<AdminProductDetail> {
  const res = await fetch(
    apiUrl(`/api/v1/admin/products/${encodeURIComponent(id)}/publish`),
    {
      method: "POST",
      credentials: "include",
      headers: { Accept: "application/json" },
    },
  );
  const data = await parseJson(res);
  if (!res.ok) throw errorFromResponse(res, data);
  return mapDetail(data);
}

/** POST /api/v1/admin/products/{id}/unpublish */
export async function unpublishAdminProduct(
  id: string,
): Promise<AdminProductDetail> {
  const res = await fetch(
    apiUrl(`/api/v1/admin/products/${encodeURIComponent(id)}/unpublish`),
    {
      method: "POST",
      credentials: "include",
      headers: { Accept: "application/json" },
    },
  );
  const data = await parseJson(res);
  if (!res.ok) throw errorFromResponse(res, data);
  return mapDetail(data);
}
